app/src/main/java/com/danosoftware/galaxyforce/services/googleplay/GooglePlayServices.java

```diff
@@ -0,0 +1,548 @@
+                    @Override
+                    public void onSuccess(SnapshotMetadata snapshotMetadata) {
+                        int wave = savedGame != null ? savedGame.getHighestWaveReached() : 0;
+                        Log.i(ACTIVITY_TAG, "Game Saved: Wave " + wave);
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        Log.e(ACTIVITY_TAG, "Game Saved Failed", e);
+                    }
+                });
+    }
+
+    /**
+     * Return a task that will asynchronously attempt to sign-in to Google Play Services.
+     */
+    private Task<GoogleSignInAccount> silentSignInTask() {
+        return signInClient.silentSignIn()
+                .addOnSuccessListener(new OnSuccessListener<GoogleSignInAccount>() {
+                    @Override
+                    public void onSuccess(GoogleSignInAccount googleSignInAccount) {
+                        Log.i(ACTIVITY_TAG, "Success signedIn");
+                        onConnected(googleSignInAccount, ConnectionRequest.LOG_IN);
+                    }
+                }).addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        Log.e(ACTIVITY_TAG, "Failed signIn", e);
+                        onDisconnected(ConnectionRequest.LOG_IN);
+                    }
+                });
+    }
+
+    /**
+     * Load the latest snapshot and resolve any conflicts (if any).
+     */
+    private Task<Snapshot> loadSnapshotAndResolveConflictsTask(GoogleSignInAccount account) {
+
+        // Get the SnapshotsClient from the signed in account.
+        SnapshotsClient snapshotsClient =
+                Games.getSnapshotsClient(mActivity, account);
+
+        // Open the saved game using its name.
+        return snapshotsClient
+                .open(SAVED_GAME_FILENAME, true)
+                .addOnFailureListener(new OnFailureListener() {
+                    @Override
+                    public void onFailure(@NonNull Exception e) {
+                        Log.e(ACTIVITY_TAG, "Error while opening Snapshot.", e);
+                    }
+                })
+                .continueWithTask(
+                        new Continuation<
+                                SnapshotsClient.DataOrConflict<Snapshot>,
+                                Task<Snapshot>>() {
+                            @Override
+                            public Task<Snapshot> then(
+                                    @NonNull Task<SnapshotsClient.DataOrConflict<Snapshot>> task)
+                                    throws Exception {
+                                return processSnapshotAndResolveConflictsTask(
+                                        task.getResult(),
+                                        MAX_SNAPSHOT_RESOLVE_RETRIES);
```

Comment:
Think this should be called initially with 0. Each recursive call increments this by 1 up to max re-tries.